### PR TITLE
Check the template...

### PR DESCRIPTION
**Pull request by maniackcrudelis**
Will merge in 3 days...

### CODE REVIEW

File: sub_scripts/testing_process.sh [hunk 1]
```diff
@@ -36,6 +36,187 @@ check_false_positive_error () {
...
+                show_error "This helper does not look like an existing helper." "Do not use specific functions aside of ynh helpers." "Keep your code linear directly into the scripts to ease the reading of your scripts"
+                break
+            fi
+        done <<< "$(echo "$helper_like")"
+    fi
+
+
+    ### Detect usage of ynh_render_template
+    template_output="$($grep_command --word-regexp "ynh_render_template" "$app_scripts_dir")"
+    show_error "Do not use ynh_render_template in app scripts." "Use ynh_replace_string instead to keep clear all modifications made on files"
```
Comment (autra) [inline sub_scripts/testing_process.sh:168]: I'm late of course, but can I know the rationale of this rule? I find ynh_render_template very useful and I don't understand why `ynh_replace_string` is safer with respect to the modifications an admin could have made to a config file for instance. The whole file is getting replaced anyway right?

Comment (maniackcrudelis) [reply sub_scripts/testing_process.sh:168]: @alexAubin I see your eyes on this comment. So sly :smiling_imp:

Too lazy to find you the comment in the core where someone tried to make me believe the helper would never be used in an app. But I knew already it would sooner or later.

If you read the error message, you'll see that it's not about modification the admin could have done on the file, but "to keep clear all modifications made on files". In other words, to keep clear what YOU change into the files in your package.

The concept is simple, if you use `ynh_replace_string`, you're going to put a call to this helper to each variable you change in the config file, or each part of the files you change for any other reasons.
When another packager read your code, or an user wondering why it doesn't work the way it should for him, it can read the code and know which part was changed and why.

When using `ynh_render_template`, you add one call that says nothing but I do some changes on the file...
Then you have to go to the file, try to untangle what's jinja2 syntax and what's the config file syntax (considering you do know that's actually jinja2 working under the hood...), spot the parts that will be actually modified by jinja2.
And, you're not yet done, you have to go back to the script to know what were the values that could have been modified, if there's some conditions, you have to find in the script about what they actually refer to.

So... either the user or packager is really willing to solve the issue, or have a lot of time to waste figuring out how the file is actually modified. Or he will just trow out the package and think it's a piece of shit...
(I would likely go for the second solution...)

Instead, with `ynh_replace_string` the user or packager will read the code, see an if with a variable defined several lines before, and will quickly see what is modified, what's not.
And why if you put a comment to explain.

So, as usual, I'd like the code to be straight, clear and simple.
so anyone can work on every app even if he doesn't know how to package an app nor even how to code.

Comment (alexAubin) [reply sub_scripts/testing_process.sh:168]: We've already had that debate a long time ago so not gonna re-trigger the entire discussion, but to give another perspective with a more nuanced pro/cons on the matter : ... basically we have our own custom "template" technique in Yunohost and it's mainly there from historical reasons.

Jinja templates (or more generally template engine, nothing specific to jinja here) are used by many people in a wide variety of contexts and are powerful and flexible. The point of such mechanism is precisely that you can focus on the template itself and not have to go back-and-forth all the time trying to understand if the code is gonna replace keyword X by value `$y` or `$z` ... In any project that regularly deals with templates, having to maintain a list of function calls, each of them replacing `foo` by content `$FOO`, another one replacing `bar` by content `$BAR` and so on... would just sound like a waste of time when you can have the template engine do that automatically for you ... In fact just debating the use of it would feel weird and it would be argued that not using a template engine is what decreases the readability / maintainability ...

Yes, template engines are a bit of magic and it's an additional abstraction to learn. Yes, it's a different syntax that needs to be learned... (well, mainly writing `{{ foo }}` instead of `__FOO__`) ... In the end, we *do* end up having our very-own custom template engine with our very-own abstractions (helpers) that handle the substitution's magic. So if you have to learn something, why not the one that a) can be reused in other projects and b) gives you more flexibility / features. Instead, using custom homemade tools will legitimately piss people who are already familiar with more standard tools and it will look as if we're doing things differently just for the sake of it...

Now in the specific context of YunoHost and from a purely practical point of view (and to paraphrase Maniack's argument which he regularly exposes) :
- Using a "simple" approach with explicit code saying « replace keyword X by value $x » is less magic for people not familiar with programming. Less abstraction = lower entry cost. 
    - (though imho somebody could argue that this argument assumes that we're not expecting people familiar with programming to contribute, or that it sounds like it's saying 'people are too dumb to understand template engine')
     - (and again, imho we have our own abstractions in the end... they just sound "less abstract" to us when we wrote them ourselves...)
- It's an historical and now de-facto practice all over the place. Having a single standard common practice is better than having X different practices with people fighting over what's the best one. Especially when you jump in a random app and want to be able to fix it quickly without taking an entire day understanding how this exact app works.
- The cost of refactorizing all apps to use another template technique is probably not worth the gain. It works and there are simply too many other more important things to work on.

So TL;DR : Yes for the "sticking to the common practice in the project increases the readability / maintainability by other people in the project", and "from a pragmatic point of view, it works and we have better things to do that rewrite everything". But I disagree with the part that seems to imply jinja or template engines in general are intrinsically bad or as if they are an intentional attempt at making things less clear...

Comment (autra) [reply sub_scripts/testing_process.sh:168]: Many thanks to both of you to have taken the time to answer me.

> it's not about modification the admin could have done on the file, but "to keep clear all modifications made on files".

Oh, you actually meant to *make clear*, as ["keep clear" has a quite different meaning](https://www.wordreference.com/enfr/keep%20clear), I didn't understand it at first :-D But let's not worry about that.

I still don't get why replacing string is easier when debugging. For me at best it doesn't change anything. When you debug, you go for the source of trust first, which is not the install script, not even the upstream repo... but your filesystem. Once you've pinpointed the error in a conf file for instance, you're going to go to `foo_ynh/conf/the_conf` and look for the values. There, it doesn't make any diff if it is `__APP__` or `{{app}}`, you're going to `git grep the_conf` on the repo and quickly find the place where it is used...

I agree with the fact that a common convention is good and that the cost of refactoring is too high. so to make things clear I don't want you guys to change the convention of course and I'm not going to argue endlessly past this comment. And I'm the first one to say that abstraction layer has a cost and shouldn't be used lightly, but forbidding them altogether, isn't it a bit extreme?

I just think that there's a big difference between having an historic convention, and automatically enforcing this convention (which is recent). The first alternative acknowledges that there are valid cases where this convention is not good/valid, the second does not. Some rules are better enforced by humans than by robots.

So even though I'm really new to the project, for the sake of flexibility and trusting app packagers to do the right thing in their context, please consider relaxing at least 2 rules in the future:

- this "template forbidden" rule, because templating and string replacement are not equivalent functionally speaking. templating is a lot safer (`__APP__` could very well be legitimately used in a upstream project conf file, and jinja has useful filters for escaping), and there's thing you simply cannot do easily with string replacement, for instance loops.
- and also the ["function forbidden" rule](https://github.com/YunoHost/package_check/pull/75/files#diff-1a3d22304b1ceafbccc50b17adf6d205R98), mainly because it prevents to use `local` variables, which can make a long install script a lot safer (avoiding var collision and improving code readability). 

I merely ask for more flexibility.

Comment (maniackcrudelis) [reply sub_scripts/testing_process.sh:168]: I wasn't talking about you debugging your installation or any user debugging its own installation, which indeed is about the files actually installed on the server.
I was talking about me and other packagers debugging the app for an user, which then is about the scripts themselves. Not the files actually installed on the server.
Prior to know what is actually installed, most importantly we want to know is what was suppose to happen.

I actually would have prefer not adding any parser in package_check to spot the usage of this helper and specific functions.
But... having to work on different packages recently... I saw that actually there's way too many packages with that kind of things.

And, having already lost a lot of hours trying to figure out how a packager made its package... I'd rather cut the problem directly at the source.
Packages change often of maintainers, so the usage of a template easy for everyone, especially others than the one who code it in the first place, become more and more important...
And packagers here seems to forgot the most important part of it, the package are for the users, not for the packagers. And so the package has to be as resilient as possible, which is mainly done by other packagers able to take over any packages at any moment.

> it prevents to use local variables

That sounds like the very bad idea of a main () and local variable everywhere...
Bash is not C.
To use local variables into the main part of a bash script is merely useless...

Considering the size of an app script, I would hardly says that it's a "long" script...


Now, no doubt we'll have side effects, already happen on several apps. I had to work on my own apps for that.
But considering some packagers are just not giving a shit about the template... We have to find solution to sort out the packages that respect the template and other that don't...
